File: aisdotnetclient/trunk/AppClient/AUtilities.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace AppClient
{
    public sealed class AUtilities
    {
        /// <summary>
        /// Converts and iEnd-separated string of name-value pairs into a map.
        /// </summary>
        /// <param name="iStr">String to be converted.</param>
        /// <param name="iMap">Holds the generated map.</param>
        /// <param name="iSep">Name-Value separator.</param>
        /// <param name="iEnd">Name-Value pair separator.</param>
        /// <returns></returns>
        public static bool StringToStringDictionary(string iStr, 
            ref SortedDictionary<string, string> iMap, string iSep, string iEnd)
        {
            if (iStr.Length == 0)
                return false;

            string[] aPairs = iStr.Split(iEnd.ToCharArray(), StringSplitOptions.None);
            string aName;
            string aValue;
            int aSize;

            aSize = aPairs.Length;

            if (iSep == iEnd)
            {
                for (int aIdx = 0; aIdx < aSize; ++aIdx)
                {
                    aName = aPairs[aIdx];
                    if (aName.Length > 0)
                    {
                        if (++aIdx != aSize)
                            aValue = aPairs[aIdx];
                        else
                            aValue = "";

                        iMap.Add(aName, aValue);
                    }
                }
            }
            else
            {
                string[] aPair = null;

                for (int aIdx = 0; aIdx < aSize; ++aIdx)
                {
                    aPair = aPairs[aIdx].Split(iSep.ToCharArray(), StringSplitOptions.None);
                    aName = aPair[0];
                    if (aName.Length != 0 && aPair.Length > 1)
                        iMap.Add(aName, aPair[1]);
                }
            }

            return true;
        }

        /// <summary>
        /// Converts and DEL-separated string of name-value pairs into a map.
        /// </summary>
        /// <param name="iStr">String to be converted.</param>
        /// <param name="iMap">Holds the generated map.</param>
        /// <returns></returns>
        public static bool StringToStringDictionary(string iStr, ref SortedDictionary<string, string> iMap)
        {
            return StringToStringDictionary(iStr, ref iMap, "\x7F", "\x7F");
        }

        /// <summary>
        /// Returns true if the specified string contains an error.
        /// </summary>
        /// <param name="iRet"></param>
        /// <returns></returns>
        public static bool ErrorReturned(string iRet)
        {
            return (iRet.Length != 0 && iRet.StartsWith("!") && iRet.EndsWith("!"));
        }

        public static int GetArgNumber(byte[] iBuffer, int iOffset, ref int irOut)
        {
            bool aIsNegative = false;
            int aDigit = 0;

            irOut = 0;

            if (iBuffer == null)
            {
                Console.Error.WriteLine("AUtilities::GetArgNumber, iBuffer is null");
                return -1;
            }

            if (iOffset < 0 || iOffset >= iBuffer.Length)
            {
                Console.Error.WriteLine("AUtilities::GetArgNumber, iOffset is invalid");
                return -1;
            }

            // skip whitespaces
            while (iOffset < iBuffer.Length && char.IsWhiteSpace((char)iBuffer[iOffset]))
                ++iOffset;

            if (iOffset == iBuffer.Length)
            {
                Console.Error.WriteLine("AUtilities::GetArgNumber, buffer is empty");
                return -1;
            }

            // check for negative number
            if ((char)iBuffer[iOffset] == '-')
            {
                aIsNegative = true;
                ++iOffset;

                if (iOffset == iBuffer.Length)
                {
                    Console.Error.WriteLine("AUtilities::GetArgNumber, incomplete");
                    return -1;
                }
            }

            if ((char)iBuffer[iOffset] <= '\x01')
            {
                return 0;
            }
            else if (!char.IsDigit((char)iBuffer[iOffset]))
            {
                Console.Error.WriteLine("AUtilities::GetArgNumber, non-numeric data");
                return -1;
            }
            else
            {
                for (; (aDigit = (char)iBuffer[iOffset] - '0') >= 0 && aDigit <= 9; ++iOffset)
                    irOut = irOut * 10 + aDigit;
            }

            if (aIsNegative)
                irOut = -irOut;

            return (iOffset + (iBuffer[iOffset] > 0x01 ? 1 : 0));
        }

        public static int GetArgString(byte[] iBuffer, char iDelimiter, int iOffset, ref string irOut)
        {
            char aC;
            int aEndOffset = 0;
            StringBuilder aString = new StringBuilder();

            if (iBuffer == null)
            {
                Console.Error.WriteLine("AUtilities::GetArgString, iBuffer is null");
                return -1;
            }

            if (iOffset < 0 || iOffset >= iBuffer.Length)
            {
                Console.Error.WriteLine("AUtilities::GetArgString, iOffset is invalid");
                return -1;
            }

            // skip whitespaces
            while (iOffset < iBuffer.Length && char.IsWhiteSpace((char)iBuffer[iOffset]))
                ++iOffset;

            if (iOffset == iBuffer.Length)
            {
                Console.Error.WriteLine("AUtilities::GetArgString, buffer is empty");
                return -1;
            }

            // scan for the separator, build output string
            for (aEndOffset = iOffset; (aC = (char)iBuffer[aEndOffset]) != iDelimiter && aC > '\x01'; ++aEndOffset)
            {
                aString.Append(aC);
            }

            irOut = aString.ToString();

            return (aEndOffset + (iBuffer[aEndOffset] > 0x01 ? 1 : 0));
        }

        public static int ByteToString(byte[] iBuffer, int iOffset, int iLength, ref string iOut)
        {
            int aEnd = iOffset + iLength;

            if (iOffset >= iBuffer.Length || aEnd > iBuffer.Length)
            {
                return -1;
            }

            StringBuilder aString = new StringBuilder();
            char aC;

            for (int i = iOffset; i < aEnd && (aC = (char)iBuffer[i]) > '\x01'; ++i)
            {
                aString.Append(aC);
            }

            iOut = aString.ToString();

            return aEnd;
        }
    }
}

File: aisjavaclient/trunk/src/com/ecwise/ais/AUtilities.java
package com.ecwise.ais;

import java.util.Hashtable;

public class AUtilities {
	
	/**
	 * Joins the elements of a String array into a single String.
	 * @param iArray
	 * @param iSeparator
	 * @return
	 */
	public static String joinStringArray(String[] iArray, String iSeparator) {
		
		StringBuilder aStringBuilder = new StringBuilder();
		
		for (String iItem : iArray) {
			aStringBuilder.append(iItem + iSeparator);
		}
		
		return aStringBuilder.toString();
	}
	
	/**
	 * Joins the elements of a String array into a single String.
	 * @param iArray
	 * @param iSeparator
	 * @param iCount
	 * @return
	 */
	public static String joinStringArray(String[] iArray, String iSeparator, int iCount) {
		
		StringBuilder aStringBuilder = new StringBuilder();
		int aCount = Math.min(iCount, iArray.length);
		
		for (int i = 0; i < aCount; i++) {
			aStringBuilder.append(iArray[i] + iSeparator);
		}
		
		return aStringBuilder.toString();
	}
	
	/**
	 * Returns true if the given string is an error message.
	 * @param iStr
	 * @return
	 */
	public static boolean isError(String iStr) {
		return (iStr != null && !iStr.isEmpty() && iStr.startsWith("!") && iStr.endsWith("!"));
	}
	
	/**
	 * Breaks a DEL-delimited string into name value pairs.
	 * @param iStr
	 * @return
	 */
	public static Hashtable<String, String> stringToStringDictionary(String iStr) {
		return stringToStringDictionary(iStr, "\177", "\177");
	}
	
	/**
	 * Breaks a string into name value pairs.
	 * @param iStr
	 * @param iSep Separator between name-value pairs. Can be the same as iEnd.
	 * @param iEnd Separator between name and value. Can be the same as iSep.
	 * @return
	 */
	public static Hashtable<String, String> stringToStringDictionary(String iStr,
			String iSep, String iEnd) {
		
		String[] aPairs = iStr.split(iSep);
		String aName;
		String aValue;
		int aSize = (aPairs != null) ? aPairs.length : 0;
		Hashtable<String, String> aDictionary = new Hashtable<String, String>();
		
		if (iSep == iEnd) {
			for (int i = 0; i < aSize; i++) {
				aName = aPairs[i];
				if (!aName.isEmpty()) {
					if (++i != aSize) {
						aValue = aPairs[i];
					} else {
						aValue = "";
					}
					aDictionary.put(aName, aValue);
				} else {
					// if the name field is empty, then skip that pair
					i+=2;
				}
			}
		} else {
			for (int i = 0; i < aSize; i++) {
				String[] aPair = aPairs[i].split(iEnd);
				if (aPair != null && (aPair.length > 1)) {
					aDictionary.put(aPair[0], aPair[1]);
				}
			}
		}
		
		return aDictionary;
	}
}
